ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,510 @@
+     */
+    if(num_args == 1) {
+        if(strcmp(arg_strings[0], "all") == 0)
+            all_on = 1;
+    }
+
+    int prev = -1;
+    /* Turn on only the counters that were specified in the MCA parameter */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        if(all_on)
+            attached_event[i] = 1;
+        else {
+            /* Note: If no arguments were given, this will be skipped */
+            for(j = 0; j < num_args && found < num_args; j++) {
+                if(strcmp(counter_names[i], arg_strings[j]) == 0) {
+                    attached_event[i] = 1;
+                    found++;
+                }
+            }
+        }
+
+        /* ########################################################################
+         * ################## Add Timer-Based Counter Enums Here ##################
+         * ########################################################################
+         */
+        /* If this is a timer event, sent the corresponding timer_event entry to 1 */
+        if(i == OMPI_MATCH_TIME)
+            timer_event[i] = 1;
+        else
+            timer_event[i] = 0;
+
+        /* Registers the current counter as an MPI_T pvar regardless of whether it's been turned on or not */
+        ret = mca_base_pvar_register("ompi", "runtime", "spc", counter_names[i], counter_descriptions[i],
+                                     OPAL_INFO_LVL_4, MPI_T_PVAR_CLASS_SIZE,
+                                     MCA_BASE_VAR_TYPE_UNSIGNED_LONG_LONG, NULL, MPI_T_BIND_NO_OBJECT,
+                                     MCA_BASE_PVAR_FLAG_READONLY | MCA_BASE_PVAR_FLAG_CONTINUOUS,
+                                     ompi_spc_get_count, NULL, ompi_spc_notify, NULL);
+
+        /* Initialize the mpi_t_indices array with the MPI_T indices.
+         * The array index indicates the SPC index, while the value indicates
+         * the MPI_T index.
+         */
+        if(ret != OPAL_ERROR) {
+            mpi_t_indices[i] = ret;
+
+            if(mpi_t_offset == -1) {
+                mpi_t_offset = ret;
+                prev = ret;
+            } else if(ret != prev + 1) {
+                mpi_t_disabled = 1;
+                fprintf(stderr, "There was an error registering SPCs as MPI_T pvars.  SPCs will be disabled for MPI_T.\n");
+                break;
+            } else {
+                prev = ret;
+            }
+        } else {
+            mpi_t_indices[i] = -1;
+            mpi_t_disabled = 1;
+            fprintf(stderr, "There was an error registering SPCs as MPI_T pvars.  SPCs will be disabled for MPI_T.\n");
+            break;
+        }
+    }
+}
+
+/* Frees any dynamically alocated OMPI SPC data structures */
+void ompi_spc_fini()
+{
+#if SPC_ENABLE == 1
+    if(!ompi_mpi_spc_dump_enabled)
+        goto skip_dump;
+
+    int i, j, rank, world_size, offset, err;
+    long long *recv_buffer, *send_buffer;
+
+    ompi_communicator_t *comm = &ompi_mpi_comm_world.comm;
```

Comment:
It may not be safe to use MPI_COMM_WORLD here.  I think you might want to dup COMM_WORLD during your startup and then use that here in the dump (and then be sure to free that communicator when you're done with it).